opengrok-indexer/src/main/java/org/opengrok/indexer/analysis/ruby/RubyLexer.java

```diff
@@ -550,39 +476,116 @@ public int nameLength(String capture) {
-        patb.append(Pattern.quote(String.valueOf(endqchar)));
-        if (nestqchar != '\0') {
-            patb.append(Pattern.quote(String.valueOf(nestqchar)));
+        patb.append(Pattern.quote(String.valueOf(d.endqchar)));
+        if (d.nestqchar != '\0') {
+            patb.append(Pattern.quote(String.valueOf(d.nestqchar)));
         }
         patb.append("]");
         patb.append(RegexUtils.getNotFollowingEscapePattern());
-        collateralCapture = Pattern.compile(patb.toString());
-        return collateralCapture;
+        d.collateralCapture = Pattern.compile(patb.toString());
+        return d.collateralCapture;
     }
 
     /**
-     * Calls {@link RubyLexer#phLOC()} if the yystate is not SCOMMENT or POD.
+     * Calls {@link #phLOC()} if the yystate is not SCOMMENT or POD.
      */
     public void chkLOC() {
-        int yystate = lexer.yystate();
-        if (yystate != SCOMMENT && yystate != POD) {
-            lexer.phLOC();
+        int yystate = yystate();
+        if (yystate != SCOMMENT() && yystate != POD()) {
+            phLOC();
         }
     }
 
+    /**
+     * Subclasses must override to possibly set the INTRA state.
+     */
+    abstract void maybeIntraState();
+
+    void pushData() {
+        if (data == null) {
+            data = new Stack<>();
+        }
+        data.push(d);
+        d = new RubyLexerData();
+    }
+
+    void popData() {
+        d = data.pop();
+    }
+
+    /**
+     * Subclasses must override to get the constant value created by JFlex to
+     * represent QUOxLxN.
+     */
+    abstract int QUOxLxN();
+
+    /**
+     * Subclasses must override to get the constant value created by JFlex to
+     * represent QUOxN.
+     */
+    abstract int QUOxN();
+
+    /**
+     * Subclasses must override to get the constant value created by JFlex to
+     * represent QUOxL.
+     */
+    abstract int QUOxL();
+
+    /**
+     * Subclasses must override to get the constant value created by JFlex to
+     * represent QUO.
+     */
+    abstract int QUO();
+
+    /**
+     * Subclasses must override to get the constant value created by JFlex to
+     * represent HEREinxN.
+     */
+    abstract int HEREinxN();
+
+    /**
+     * Subclasses must override to get the constant value created by JFlex to
+     * represent HERExN.
+     */
+    abstract int HERExN();
+
+    /**
+     * Subclasses must override to get the constant value created by JFlex to
+     * represent HEREin.
+     */
+    abstract int HEREin();
+
+    /**
+     * Subclasses must override to get the constant value created by JFlex to
+     * represent HERE.
+     */
+    abstract int HERE();
+
+    /**
+     * Subclasses must override to get the constant value created by JFlex to
+     * represent SCOMMENT.
+     */
+    abstract int SCOMMENT();
+
+    /**
+     * Subclasses must override to get the constant value created by JFlex to
+     * represent POD.
+     */
+    abstract int POD();
+
     private class HereDocSettings {
```

Comment:
this class is the same like in PerlAnalyser, can you extract it so the code is not duplicated?

Reply:
This is a parallel-evolved feature in the languages but not always the same; and cross-language, inter-analyzer dependencies are hard to reason about for the limited value of re-use.

Also the similar features have to be allowed to change behavior independently.

E.g. support for indented here-docs (partially but not 100% supported by OpenGrok for Ruby or Perl) arrived in Ruby years before Perl's, so `HereDocSettings` would have evolved differently from Perl's.

Reply:
Ok, I agree for separating the analyzers logic.

It's still violating the YAGNI principle but for this instance it makes more sense to keep it separately.